specs/emip-005.md

```diff
@@ -0,0 +1,472 @@
+Ed25519PubKey = 2
+
+EncryptionType : uint = ChachaPoly / Ed25519PubKey
+```
+
+We will explain what this means shortly but you can find the full list of the types of encryption allowed by COSE at [RFC 8152 section 5.1.1](https://tools.ietf.org/html/rfc8152#section-5.1.1)
+
+### Structure
+
+The COSE specification is made to be composable -- that is you can have a plaintext that you wrap with a signature, then wrap with an encryption, then wrap with a signature again (and so on).
+
+That means that for encryption in particular, it can either
+
+1) Be used to encrypt plaintext directly
+2) Be used to encrypt another COSE message
+
+In this spec, we care about the case where you encrypt a `signed_message`
+
+Here is the overall CBOR structure
+
+```
+# holds encrypted content itself
+COSE_Encrypt = [
+  Headers,
+  ciphertext : bstr / nil, # contains encrypted signed_message
+  recipients : [+COSE_recipient]
+]
+
+# holds encrypted keys the receiver can use to decrypt the content
+COSE_recipient = [
+  Headers,
+  ciphertext : bstr / nil, # contains encrypted key to decrypt the COSE_Encrypt ciphertext
+  ? recipients : [+COSE_recipient] # in case you need multiple rounds to get decryption key
+]
+```
+
+To encrypt, the structure as a whole, we call our encryption method once for each level (root, recipient, etc.) recursively. For example, we encrypt the `signed_message` and put it in the `COSE_Encrypt` ciphertext, then we encrypt the decryption key and put it in the `COSE_recipient` ciphertext.
+
+For the `Headers`,
+
+- The `protected` fields MUST be empty. These are meant to be used with AEAD which we  don't need in this  specification (you can read more about it at [RFC 8152 section 5.3](https://tools.ietf.org/html/rfc8152#section-5.3)).
+- The `unprotected` field MUST contain a `enc_type` label with value `EncryptionType` so that wallets can easily know what UI to display to users.
+
+We define the encrypted payload as the union of two possibilities (defined below)
+
+```
+encrypted_message = PasswordEncryption / PubKeyEncryption
+```
+
+#### Password-based encryption
+
+For password-based encryption we don't need a receiver field (anybody who knows the password can decrypt) so we instead use the following (simplified) structure
+
+```
+COSE_Encrypt0 = [
+    Headers,
+    ciphertext : bstr / nil,
+]
+
+PasswordEncryption = 16 (COSE_Encrypt0)
+```
+
+The `Headers` MUST have `enc_type: 1` in the `unprotected` field.
+
+The `ciphertext` is simply the result of [the Yoroi encryption spec](https://github.com/Emurgo/yoroi-frontend/blob/737595fec5a89409aacef827d356c9a1605515c0/docs/specs/code/ENCRYPT.md) on the `signed_message`.
+
+#### Public key based encryption
+
+We only allow encrypting based on `ED25519 extended` public keys (the ones used for Cardano). To encrypt based on these public keys, you must
+
+1) Compute a password consisting of 22 case-sensitive alphanumeric (a–z, A–Z, 0–9) characters (this gives you ~128 bits of entropy)
```

Comment:
Using this technique allows us to encrypt a message for multiple receivers but it's slightly more bytes than needed for just a single receiver (which is probably the more common case). 

We could have a single-receiver pubkey-based encryption mode where it uses the `COSE_Encrypt0` format with the hash of the shared secret as the password.